server/src/Utils/Order.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Utils.Order where
+
+import Import.NoFoundation
+import Model.Types
+
+-- {-| Determine if a given user is allowed to access the PDF file of an item.
+-- -}
+hasAccessToPdfFileDownload ::
+       ( BaseBackend (YesodPersistBackend site) ~ SqlBackend
+       , YesodPersist site
+       , PersistQueryRead (YesodPersistBackend site)
+       )
+    => UserId
+    -> Text
+    -> HandlerT site IO AuthResult
+hasAccessToPdfFileDownload userId filename = do
+    let unauthorized = Unauthorized "You didn't buy this item"
```

Comment:
Ok. I tried out the monad transformer approach just now and almost got it to compile! Basically you just wrap all your `HandlerT ...` monads in `MaybeT`, as in `(Entity pdfId _) <- MaybeT $ runDB $ selectFirst [PdfFileFilename ==. filename] []`